### PR TITLE
Thunderbolt fixes

### PR DESCRIPTION
**Pull request by westeri**
This pull request contains fixes for review comments after the Titan Ridge support patch series was merged.

There is one additional commit that makes the Thunderbolt plugin to parse nvm_version everywhere in the same way.

### CODE REVIEW

**Comment by westeri** [comment]
@YehezkelShB, these should fix your findings. Could you please review?

File: plugins/thunderbolt/fu-thunderbolt-image.c [hunk 1]
```diff
@@ -491,24 +491,77 @@ get_host_locations (guint16 id)
...
+	if (fw_mc_loc.offset == 0)
+		return TRUE;
+
+	if (!read_drom_entry_location (image, DROM_ENTRY_MC, &image_mc_loc, error))
+		return FALSE;
+
+	if (image_mc_loc.offset == 0) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
+				     "The FW image file misses multi controller entry");
```
Comment (hughsie) [inline plugins/thunderbolt/fu-thunderbolt-image.c]: drop the `The` -- perhaps something like `firmware does not have multi-controller entry point`

File: plugins/thunderbolt/fu-thunderbolt-image.c [hunk 2]
```diff
@@ -491,24 +491,77 @@ get_host_locations (guint16 id)
...
+		return FALSE;
+	}
+	if (fw_mc_loc.len != image_mc_loc.len) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
+				     "The FW image multi controller entry length mismatch");
```
Comment (hughsie) [inline plugins/thunderbolt/fu-thunderbolt-image.c]: same

File: plugins/thunderbolt/fu-thunderbolt-image.c [hunk 3]
```diff
@@ -491,24 +491,77 @@ get_host_locations (guint16 id)
...
+	if (fw_mc == NULL)
+		return FALSE;
+	image_mc = read_location (&image_mc_loc, image, error);
+	if (image_mc == NULL)
+		return FALSE;
+
+	if (memcmp (fw_mc->data, image_mc->data, fw_mc->len) != 0) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
+				     "The FW image multi controller entry mismatch");
```
Comment (hughsie) [inline plugins/thunderbolt/fu-thunderbolt-image.c]: same

File: plugins/thunderbolt/fu-thunderbolt-image.c [hunk 4]
```diff
@@ -520,11 +573,9 @@ get_device_locations (guint16 id, const FuThunderboltFwObject *fw, GError **erro
 	case 0x15EA:
 	case 0x15EF:
 		/* if the controller has multi controller entry need to
-		 * compare it against the image. */
-		if (read_drom_entry_location (fw, DROM_ENTRY_MC, &TR[1],
-					      &error_local))
+		 * compare it against the image first. */
+		if (compare_device_mc (fw, image, error))
```
Comment (hughsie) [inline plugins/thunderbolt/fu-thunderbolt-image.c]: You need the error_local and g_deubg perhaps? Either you return VALIDATION_FAILED and set error, or return TR and don't touch error. If that's what's happening, it's slightly odd to rely on the fallthrough IMHO.

Comment (westeri) [reply plugins/thunderbolt/fu-thunderbolt-image.c]: Hmm, so the idea is to return VALIDATION_FAILED and propagate error if the `compare_device_mc ()` fails. Otherwise return TR. I wonder if this would be better than the fallthrough case:
```
if (compare_device_mc (fw, image, error))
        return TR;
return NULL;
```
?

**Comment by westeri** [comment]
Looks like self tests are using "raw" nvm_version such as "20.0" and the patch included in this pull request makes it fail because it converts all nvm_version strings to the fwupd format "20.00". I suppose here the correct fix is to update the self tests accordingly? @superm1?

File: plugins/thunderbolt/fu-thunderbolt-image.c [hunk 5]
```diff
@@ -491,24 +491,77 @@ get_host_locations (guint16 id)
...
+ * controller and image MC entries match. In any other case FALSE is
+ * returned and error is set accordingly.
+ */
+static gboolean
+compare_device_mc (const FuThunderboltFwObject *fw,
+		   const FuThunderboltFwObject *image,
+		   GError **error)
+{
+	FuThunderboltFwLocation image_mc_loc = { .section = DROM_SECTION, .description = "Multi Controller" };
+	FuThunderboltFwLocation fw_mc_loc = { .section = DROM_SECTION, .description = "Multi Controller" };
```
Comment (YehezkelShB) [inline plugins/thunderbolt/fu-thunderbolt-image.c]: Matter of taste, but some would prefer `= image_mc_loc` instead of repeating the literal info.

Comment (YehezkelShB) [reply plugins/thunderbolt/fu-thunderbolt-image.c]: Also, I think most of the code uses `controller`, as `fw` is a bit confusing (it can be the FW image).

Comment (YehezkelShB) [reply plugins/thunderbolt/fu-thunderbolt-image.c]: (It makes sense to change the parameter name of `get_device_locations()` too as now it gets two different `FuThunderboltFwObject` parameters)

**Comment by westeri** [comment]
I updated the branch according to your comments. I also updated self tests to include `nvm_parsed_version` that should match the version returned by fwupd functions.

File: plugins/thunderbolt/fu-plugin-thunderbolt.c [hunk 6]
```diff
@@ -174,6 +174,20 @@ fu_plugin_thunderbolt_find_nvmem (GUdevDevice  *udevice,
...
+		return NULL;
+	split = g_strsplit (version_raw, ".", -1);
+	if (g_strv_length (split) != 2)
+		return NULL;
+	return g_strdup_printf ("%02x.%02x",
+				(guint) g_ascii_strtoull (split[0], NULL, 16),
+				(guint) g_ascii_strtoull (split[1], NULL, 16));
+}
+
 static const gchar *
```
Comment (hughsie) [inline plugins/thunderbolt/fu-plugin-thunderbolt.c]: this can't be const if it's returning an allocated string

**Comment by westeri** [comment]
Hopefully fixed the build issue now.

File: plugins/thunderbolt/fu-self-test.c [hunk 7]
```diff
@@ -1174,7 +1180,7 @@ test_update_fail (ThunderboltTest *tt, gconstpointer user_data)
 	/* version should *not* have changed */
```
Comment (superm1) [inline plugins/thunderbolt/fu-self-test.c]: I think this comment should actually be updated since the version is changing.

File: plugins/thunderbolt/fu-self-test.c [hunk 8]
```diff
@@ -792,18 +793,21 @@ static MockDevice root_one = {
 	.name = "Laptop",
 	.id = "0x23",
 	.nvm_version = "20.0",
+	.nvm_parsed_version = "20.00",
```
Comment (superm1) [inline plugins/thunderbolt/fu-self-test.c]: With this change, I wonder if it's also adding a device that changes in potentially non-obvious way to self tests since it's not obvious immediately that's what's happening.
eg `.nvm_version = 2.2` -> `.nvm_parsed_version = 02.02`

File: plugins/thunderbolt/fu-plugin-thunderbolt.c [hunk 9]
```diff
@@ -413,7 +411,7 @@ static void
 fu_plugin_thunderbolt_change (FuPlugin *plugin, GUdevDevice *device)
 {
 	FuDevice *dev;
-	const gchar *version;
+	g_autofree gchar *version;
```
Comment (superm1) [inline plugins/thunderbolt/fu-plugin-thunderbolt.c]: missing `= NULL`

File: plugins/thunderbolt/fu-plugin-thunderbolt.c [hunk 10]
```diff
@@ -616,7 +614,7 @@ on_wait_for_device_added (FuPlugin    *plugin,
 	FuDevice  *dev;
 	const gchar *uuid;
 	const gchar *path;
-	const gchar *version;
+	g_autofree gchar *version;
```
Comment (superm1) [inline plugins/thunderbolt/fu-plugin-thunderbolt.c]: missing `= NULL;`

File: plugins/thunderbolt/fu-self-test.c (hunk 8 above)
Comment (westeri) [reply plugins/thunderbolt/fu-self-test.c]: Not sure I follow. `nvm_parsed_version` is the expected version string after fwupd has added the corresponding device. It uses the format `%02X.%02X` so 2.2 would be 02.02. But that's exactly what happens with real devices, no?

File: plugins/thunderbolt/fu-plugin-thunderbolt.c (hunk 9 above)
Comment (westeri) [reply plugins/thunderbolt/fu-plugin-thunderbolt.c]: Oops. Not used to these g_auto* things yet.

I don't see these warnings when I run:
```
ninja -C build
```
Am I missing something?

File: plugins/thunderbolt/fu-self-test.c (hunk 8 above)
Comment (superm1) [reply plugins/thunderbolt/fu-self-test.c]: Yes, but it's not an obvious behavior.  At least when you're given a string `X.Y` but `X` and `Y` are one digit usually you treat it as `0X.0Y`.  So when the interface from the kernel gives `23.5` unless you've been watching closely you won't know that's major 23 and minor 5 and might read it as `23.50`.

I just want to make sure that nuance is captured in the self tests as the people who aren't us who look at this code down the road might not realize that's what happening.

So I think either adding another device to self tests or changing one of them to have a non-zero minor would be good.

File: plugins/thunderbolt/fu-plugin-thunderbolt.c (hunk 9 above)
Comment (superm1) [reply plugins/thunderbolt/fu-plugin-thunderbolt.c]: Looks like CI is showing them failing too.  CI runs with `-Werror`, see if they show up from that.

File: plugins/thunderbolt/fu-self-test.c (hunk 8 above)
Comment (westeri) [reply plugins/thunderbolt/fu-self-test.c]: OK, understood now.

**Comment by westeri** [comment]
I fixed the issues CI reported and changed version string of one device to be 20.2 which parses to 20.02.

**Comment by superm1** [comment]
I didn't test it this particular PR yet but I did test Mika's earlier one that already merged.

File: plugins/thunderbolt/fu-plugin-thunderbolt.c (hunk 6 above)
Comment (YehezkelShB) [reply plugins/thunderbolt/fu-plugin-thunderbolt.c]: @hughsie
"Can't" technically or by convention?

**Comment by YehezkelShB** [comment]
(But I have no way to test it. Sorry.)
